### PR TITLE
First commit

### PR DESCRIPTION
**Pull request by jaiminpanchal27**
First commit. 
- Includes gulp setup 
- Creative.js

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+	"presets": ["env"],
```
Comment (mkendall07) [inline .babelrc]: from the docs:` babel-preset-env behaves exactly the same as babel-preset-latest (or babel-preset-es2015, babel-preset-es2016, and babel-preset-es2017 together).`
Is this what we want?

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+});
+
+gulp.task('build-dev', () => {
+  return gulp.src(['src/creative.js'])
+    .pipe(webpackStream(webpackConfig))
+    .pipe(gulp.dest('build'));
+});
+
+gulp.task('build-prod', ['clean'], () => {
+  var cloned = _.cloneDeep(webpackConfig);
```
Comment (mkendall07) [inline gulpfile.js]: why do you need to clone this?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "prebid-universal-creative",
+  "version": "0.1.0",
+  "description": "",
+  "main": "index.js",
```
Comment (mkendall07) [inline package.json]: this should be dist/creative.js I think

File: src/creative.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+import * as utils from './utils';
+import * as environment from './environment';
+
+const pbjs = window.pbjs = (window.pbjs || {});
+
+/**
+ * @param  {object} doc
+ * @param  {string} adId
+ * @param  {object} dataObject
```
Comment (mkendall07) [inline src/creative.js]: let's break out the jsDoc for dataObject

File: src/creative.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+ */
+pbjs.renderAd = function(doc, adId, dataObject) {
+  if (environment.isAmp(dataObject)) {
+    renderAmpAd(dataObject.host, dataObject.uuid);
+  } else if (environment.isCrossDomain()) {
+    renderCrossDomain(adId, dataObject.pubUrl);
+  } else {
+    // assume legacy?
```
Comment (mkendall07) [inline src/creative.js]: remove comment

File: src/creative.js [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      }
+    }
+  }
+}
+
+function renderCrossDomain(adId, pubUrl) {
+  let urlParser = document.createElement('a');
+  urlParser.href = pubUrl;
+  let publisherDomain = urlParser.protocol + '//' + urlParser.host;
+  let adServerDomain = urlParser.protocol + '//tpc.googlesyndication.com';
```
Comment (mkendall07) [inline src/creative.js]: `tpc.googlesyndication.com` needs to be configurable some how

File: src/creative.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+      if (adObject.mediaType === 'video') {
+        console.log('Error trying to write ad.');
+      } else if (ad) {
+        let frame = utils.getEmptyIframe();
+        body.appendChild(frame);
+        frame.contentDocument.open();
+        frame.contentDocument.write(ad);
+        frame.contentDocument.close();
+      } else if (url) {
+        body.insertAdjacentHTML('beforeend', '<IFRAME SRC="' + url + '" FRAMEBORDER="0" SCROLLING="no" MARGINHEIGHT="0" MARGINWIDTH="0" TOPMARGIN="0" LEFTMARGIN="0" ALLOWTRANSPARENCY="true" WIDTH="' + width + '" HEIGHT="' + height + '"></IFRAME>');
```
Comment (mkendall07) [inline src/creative.js]: we should use dom manipulation to prevent XSS.  See prebid.js => `renderAd` function for example

File: src/creative.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  // TODO pass in /path from creative since it might change
+  let adUrl = 'https://' + cacheHost + '/pbc/v1/cache?uuid=' + uuid;
+  
+  
+  let handler = function(response) {
+    let bidObject;
+    try {
+      bidObject = JSON.parse(response);
+    } catch (error) {
+      // Invalid json
```
Comment (mkendall07) [inline src/creative.js]: log error?

File: src/creative.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    } else if (bidObject.nurl) {
+      let nurl = bidObject.nurl;
+      // TODO test height and width. ortb spec represents width and height as w and h.
+      utils.writeAdUrl(nurl, bidObject.h, bidObject.w);
+    }
+  };
+  utils.sendRequest(adUrl, handler);
+}
+
+// function render() {
```
Comment (mkendall07) [inline src/creative.js]: Drop dead code

File: src/environment.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+ *  DfpSafeFrame: An iframe that can't get to the top window
+ *  Sandboxed: An iframe that can't get to the top window
+ *  SuperSandboxed: An iframe without allow-same-origin
+ *  Unknown: A default sandboxed implementation delivered by EnvironmentDispatch when all positive environment checks fail
+ */
+
+/**
+ * @returns true if we are running on the top window at dispatch time
+ */
+function isCodeOnPage() {
```
Comment (mkendall07) [inline src/environment.js]: rename to something like `isTopWindow`

File: src/environment.js [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    return isSafeFrame() && href.indexOf('google') !== -1 && href.indexOf('safeframe') !== -1;
+  }
+  return false;
+}
+
+/**
+ * Return true if we are in an iframe and can't access the top window.
+ */
+export function isCrossDomain() {
+  return window.top !== window && !window.frameElement;
```
Comment (mkendall07) [inline src/environment.js]: does this throw an exception?

File: src/environment.js [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+}
+  
+/**
+ * Return true if we cannot document.write to a child iframe (this implies no allow-same-origin)
+ */
+function isSuperSandboxedIframe() {
```
Comment (mkendall07) [inline src/environment.js]: unused function

File: src/utils.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
+export function createTrackPixelHtml(url) {
+  if (!url) {
+    return '';
+  }
+
+  var escapedUrl = encodeURI(url);
+  var img = '<div style="position:absolute;left:0px;top:0px;visibility:hidden;">';
+  img += '<img src="' + escapedUrl + '"></div>';
```
Comment (mkendall07) [inline src/utils.js]: could use string templates here

File: gulpfile.js (hunk 2 above)
Comment (jaiminpanchal27) [reply gulpfile.js]: For prod we remove source-maps from webpack config. So we clone the object and do not modify original webpack config.

File: .babelrc (hunk 1 above)
Comment (jaiminpanchal27) [reply .babelrc]: They have deprecated `babel-preset-es2015` and advised to use `babel-preset-env`

File: src/environment.js (hunk 11 above)
Comment (jaiminpanchal27) [reply src/environment.js]: No it won't throw exception.

File: src/environment.js (hunk 12 above)
Comment (jaiminpanchal27) [reply src/environment.js]: removed

**Comment by jaiminpanchal27** [comment]
@mkendall07 Updated the code. We can merge this one and create separate PR for adding test framework.

File: .babelrc (hunk 1 above)
Comment (mkendall07) [reply .babelrc]: ok but what's that do to the file size?

**Pull request merged**